## 1 Д/з


# B. Кольцевая линия метро

Витя работает недалеко от одной из станций кольцевой линии Московского метро, а живет рядом с другой станцией той же линии. Требуется выяснить, мимо какого наименьшего количества промежуточных станций необходимо проехать Вите по кольцу, чтобы добраться с работы домой.

Формат ввода

Станции пронумерованы подряд натуральными числами 1, 2, 3, …, N (1-я станция – соседняя с N-й), N не превосходит 100.

Вводятся три числа: сначала N – общее количество станций кольцевой линии, а затем i и j – номера станции, на которой Витя садится, и станции, на которой он должен выйти. Числа i и j не совпадают. Все числа разделены пробелом.
Формат вывода

Требуется выдать минимальное количество промежуточных станций (не считая станции посадки и высадки), которые необходимо проехать Вите. 

In [ ]:
N, i, j = map(int, input().split())

print(min(N-abs(i-j), abs(i-j))-1)


# C. Даты



 Как известно, два наиболее распространённых формата записи даты — это европейский (сначала день, потом месяц, потом год) и американски (сначала месяц, потом день, потом год). Системный администратор поменял дату на одном из бэкапов и сейчас хочет вернуть дату обратно. Но он не проверил, в каком формате дата используется в системе. Может ли он обойтись без этой информации?

Иначе говоря, вам даётся запись некоторой корректной даты. Требуется выяснить, однозначно ли по этой записи определяется дата даже без дополнительной информации о формате.
Формат ввода
Первая строка входных данных содержит три целых числа — x, y и z (1≤x≤31, 1≤y≤31, 1970≤z≤2069. Гарантируется, что хотя бы в одном формате запись xyz задаёт корректную дату.
Формат вывода
Выведите 1, если дата определяется однозначно, и 0 в противном случае. 

In [ ]:
x, y, z = map(int, input().split())

if (x<12 or y < 12) and x == y:
  print(1)
  print()
elif x>= 12 or y>=12:
  print(1)
  print()
else:
  print(0)
  print()

In [ ]:
x, y, z = map(int, input().split())

if (x< 12 and y<12) and x != y:
  print(0)
else:
  print(1)

9 9 2026
1


# E. Точка и треугольник



На координатной плоскости расположены равнобедренный прямоугольный треугольник ABC с длиной катета d и точка X. Катеты треугольника лежат на осях координат, а вершины расположены в точках: A (0,0), B (d,0), C (0,d).

Напишите программу, которая определяет взаимное расположение точки X и треугольника. Если точка X расположена внутри или на сторонах треугольника, выведите 0. Если же точка находится вне треугольника, выведите номер ближайшей к ней вершины.
Формат ввода

Сначала вводится натуральное число d (не превосходящее 1000), а затем координаты точки X – два целых числа из диапазона от ­–1000 до 1000.

Формат вывода

Если точка лежит внутри, на стороне треугольника или совпадает с одной из вершин, то выведите число 0. Если точка лежит вне треугольника, то выведите номер вершины треугольника, к которой она расположена ближе всего (1 – к вершине A, 2 – к B, 3 – к C). Если точка расположена на одинаковом расстоянии от двух вершин, выведите ту вершину, номер которой меньше. 

In [ ]:
d = (int(input())) 
x, y = map(int, input().split())

if (x < 0 and y < 0) or (d/2 == x and y<0):
  print(1)
elif (x == 0 and d <= y) or (y == 0 and d <= x) or (y <= -x + d and x>0 and y>0):
  print(0)
elif (x < 0 and d/2 > y) or (y > -x + d) or (d/2 == y and x<0):
  print(2)
else:
  print(3)

1000
-800 900
2


In [ ]:
private static int getAnswer(int d, int x, int y) {
        if (x >= 0 && y >= 0 && x + y <= d) return 0;

        if (2 * Math.max(x, y) <= d) return 1;
        return x >= y ? 2 : 3;
    }

## 2 Д/з
# B. Дома и магазины



На Новом проспекте построили подряд 10 зданий. Каждое здание может быть либо жилым домом, либо магазином, либо офисным зданием.

Но оказалось, что жителям некоторых домов на Новом проспекте слишком далеко приходится идти до ближайшего магазина. Для разработки плана развития общественного транспорта на Новом проспекте мэр города попросил вас выяснить, какое же наибольшее расстояние приходится преодолевать жителям Нового проспекта, чтобы дойти от своего дома до ближайшего магазина.
Формат ввода


Программа получает на вход десять чисел, разделенных пробелами. Каждое число задает тип здания на Новом проспекте: число 1 обозначает жилой дом, число 2 обозначает магазин, число 0 обозначает офисное здание. Гарантируется, что на Новом проспекте есть хотя бы один жилой дом и хотя бы один магазин.

Формат вывода

Выведите одно целое число: наибольшее расстояние от дома до ближайшего к нему магазина. Расстояние между двумя соседними домами считается равным 1 (то есть если два дома стоят рядом, то между ними расстояние 1, если между двумя домами есть еще один дом, то расстояние между ними равно 2 и т.д.) 

 ** N*N **

In [ ]:
A = [int(i) for i in input().split()]
dl = []
intwo = [key for key, i in enumerate(A) if i == 2]
for i in range(len(A)):
  if A[i] == 1:
    kv = []
    for j in intwo:
      kv.append(abs(i-j))
    dl.append(min(kv))
print(max(dl))

**2*N**

In [ ]:
d = list(map(int, input().split()))
left = [0] * len(d)
shoppos = -20

for i in range(len(d)):
  if d[i] == 2:
    shoppos = i
  if d[i] == 1:
    left[i] = i - shoppos
ans = 0
shoppos = 30
for i in range(len(d)-1, -1, -1):
  if d[i] == 2:
    shoppos = i
  if d[i] == 1:
    mindist = min(shoppos - i, left[i])
    ans = max(ans, mindist)
print(ans)


2 0 1 1 0 1 0 2 1 2
3


# D. Лавочки в атриуме



В атриуме нового корпуса ФНК урбанисты установили модные гранитные лавочки (на которых холодно сидеть зимой и жарко летом). Лавочка устроена следующим образом: несколько одинаковых кубических гранитных блоков ставятся в ряд, а на них кладется гранитная плита.

При этом блоки располагаются так, чтобы плита не падала: для этого достаточно, чтобы и слева, и справа от центра плиты был хотя бы один гранитный блок или его часть (в частности, если центр плиты приходится на середину какого-нибудь блока, то и слева, и справа от центра плиты находится часть блока, и плита не падает).

На ФНК много певокурсников (но это только пока не произошли отчисления за списывания на курсе ОиМП) и им не хватает стульев в аудиториях. Студенты обнаружили, что блоки можно использовать в аудитории в качестве сиденья. Можно по одному вытаскивать блоки, находящиеся с краю (как слева, так и справа). Они хотят вытащить из-под лавочки как можно больше блоков так, чтобы она при этом не упала (передвигать оставшиеся блоки нельзя). Определите, какие блоки они должны оставить.
Формат ввода

В первой строке входных данных содержатся два числа: L - длина лавочки и K - количество гранитных блоков-ножек. Оба числа натуральные и не превышают 10 000.

Во второй строке следуют K различных целых неотрицательных чисел, задающих положение каждой ножки. Положение ножки определяется расстоянием от левого края плиты до левого края ножки (ножка - это куб размером 1×1×1). Ножки перечислены слева направо (то есть начиная с ножки с меньшим расстоянием до левого края).
Формат вывода

Требуется перечислить ножки, которые студентам нужно оставить. Для каждой ножки нужно выдать ее положение, как оно задано во входных данных. Ножки следует перечислять слева направо, в том порядке, в котором они встречаются во входных данных. 

In [ ]:
import math
dl_kol = [*map(int, input().split())]
polo = [*map(int, input().split())]
chntr = dl_kol[0] / 2
if chntr % 2 == 0.5 and (math.ceil(chntr)-1) in polo:
  print(math.ceil(chntr)-1)
else:
  chntr = math.ceil(dl_kol[0] / 2)
  print(f'центр = {chntr}')

  polo_ri, polo_le = [], []
  for i in polo:
    if i < chntr:
      polo_le.append(i)
    else:
      polo_ri.append(i)
  print(polo_le, polo_ri)
  for i in range(len(polo_le)):
    if math.ceil((polo_le[-(i+1)] + polo_ri[i]+ 2) / 2) == chntr:
      print(polo_le[-(i+1)], polo_ri[i])


14 6
1 6 8 11 12 13
центр = 7
[1, 6] [8, 11, 12, 13]
1 11


In [ ]:
import math
dl_kol = [*map(int, input().split())]
polo = [*map(int, input().split())]
chntr = dl_kol[0] / 2
if chntr % 2 == 0.5 and (math.ceil(chntr)-1) in polo:
  print(math.ceil(chntr)-1)
else:
  chntr = math.ceil(dl_kol[0] / 2)
  print(f'центр = {chntr}')

  polo_ri, polo_le = [], []
  for i in polo:
    if i < chntr:
      polo_le.append(i)
    else:
      polo_ri.append(i)
  print(polo_le, polo_ri)
  for i in range(len(polo_le)):
    if math.ceil((polo_le[-(i+1)] + polo_ri[i]+ 2) / 2) == chntr:
      print(polo_le[-(i+1)], polo_ri[i])

13 
1 4 8 11
центр = 7
[1, 4] [8, 11]
4 8
1 11


# 3 Д/з

**Количество совпадающих**

Даны два списка чисел, которые могут содержать до 100000 чисел каждый. Посчитайте, сколько чисел содержится одновременно как в первом списке, так и во втором. Примечание. Эту задачу на Питоне можно решить в одну строчку. 

In [ ]:
st1 = [*map(int, input().split())]
st2 = [*map(int, input().split())]
count = 0
for i in st1:
  if i in st2:
    count +=1
print(count)

3 4
3 3
1


In [ ]:
print(len(set(input().split()) & set(input().split())))

4 5
5 44
1


**Встречалось ли число раньше**

Во входной строке записана последовательность чисел через пробел. Для каждого числа выведите слово YES (в отдельной строке), если это число ранее встречалось в последовательности или NO, если не встречалось. 

In [ ]:
st1 = [*map(int, input().split())]

print('NO')
for i in range(1, len(st1)):
  if st1[i] in st1[:i]:
    print('YES')
  else:
    print('NO')

1 2 3 2 3 4
NO
NO 1
NO 2
YES 3 [1, 2, 3]
YES 4 [1, 2, 3, 2]
NO 5


In [ ]:
a = list(map(int, input().split()))
newset = set()
for i in a:
    print('YES' if i in newset else 'NO')
    newset.add(i)

# C. Уникальные элементы


Дан список. Выведите те его элементы, которые встречаются в списке только один раз. Элементы нужно выводить в том порядке, в котором они встречаются в списке. 

In [ ]:
a = list(map(int, input().split()))
newset = []
for i in a:
    if i not in newset:
        newset.append(i)
print(*newset)

1 2 2 3 3 3
1 2 3


In [ ]:
import collections
a=list(map(int, input().split()))
b=collections.Counter(a)
print(b)
c=[i[0]  for i  in b.items() if i[1]==1]
print(*c)

1 2 2 3 3 3
Counter({3: 3, 2: 2, 1: 1})
1


# D. Угадай число



Август и Беатриса играют в игру. Август загадал натуральное число от 1 до n. Беатриса пытается угадать это число, для этого она называет некоторые множества натуральных чисел. Август отвечает Беатрисе YES, если среди названных ей чисел есть задуманное или NO в противном случае. После нескольких заданных вопросов Беатриса запуталась в том, какие вопросы она задавала и какие ответы получила и просит вас помочь ей определить, какие числа мог задумать Август.
Формат ввода

Первая строка входных данных содержит число n — наибольшее число, которое мог загадать Август. Далее идут строки, содержащие вопросы Беатрисы. Каждая строка представляет собой набор чисел, разделенных пробелами. После каждой строки с вопросом идет ответ Августа: YES или NO. Наконец, последняя строка входных данных содержит одно слово HELP. 


Пример 1
 *Ввод*
10
1 2 3 4 5
YES
2 4 6 8 10
NO
HELP


In [ ]:
n = int(input())
set_n = set(range(1, n+1))
a=set(map(int, input().split()))
b = input()
while b != 'HELP':
  if b == 'YES':
    set_n -= a
  a = [input().split()]
  if a[0][0] != 'HELP':
    a=set(map(int, a[0][0]))
  else:
    break
  
print(n)


работает!

In [ ]:
n = int(input())
possible = set(range(1, n+1))

s = input()
while s != 'HELP':
  nums = set(map(int, s.split()))
  s = input()
  if s == 'YES':
    possible.intersection_update(nums)
  else:
    possible.difference_update(nums)
  s = input()
print(*sorted(possible))

# E. Автомобильные номера



Неизвестный водитель совершил ДТП и скрылся с места происшествия. Полиция опрашивает свидетелей. Каждый из них говорит, что запомнил какие-то буквы и цифры номера. Но при этом свидетели не помнят порядок этих цифр и букв. Полиция хочет проверить несколько подозреваемых автомобилей. Будем говорить, что номер согласуется с показанием свидетеля, если все символы, которые назвал свидетель, присутствуют в этом номере (не важно, сколько раз).
Формат ввода

Сначала задано число - количество свидетелей. Далее идет M строк, каждая из которых описывает показания очередного свидетеля. Эти строки непустые и состоят из не более чем 20 символов. Каждый символ в строке - либо цифра, либо заглавная латинская буква, причём символы могут повторяться.

Затем идёт число - количество номеров. Следующие строки представляют из себя номера подозреваемых машин и имеют такой же формат, как и показания свидетелей.
Формат вывода

Выпишите номера автомобилей, согласующиеся с максимальным количеством свидетелей. Если таких номеров несколько, то выведите их в том же порядке, в котором они были заданы на входе. 

3
ABC
A37
BCDA
2
A317BD
B137AC

In [ ]:
wits = []
for _ in range(int(input())):
  wits.append(set(input().strip()))

nums = []
maxwitcht = 0
for _ in range(int(input())):
  num = input().strip()
  numset = set(num)
  witcnt = 0
  for wit in wits:
    if wit <= numset:
      witcnt += 1
  nums.append((num, witcnt))
  maxwitcht = max(maxwitcht, witcnt)
ans = []
for num in nums:
  if num[1] == maxwitcht:
    ans.append(num[0])
print('\n'.join(ans))

# 4 Д/з

# A. Толя-Карп и новый набор структур, часть 2



Толя-Карп запросил для себя n посылок с «Аллигатор-экспресс».

Посылка представляет из себя ящик. Внутри ящика лежит целое число ai. Номер на ящике di указывает на цвет числа, лежащего внутри.

Толю-Карпа интересует, чему будут равны значения чисел, если сложить между собой все те, что имеют одинаковый цвет. Напишите, пожалуйста, программу, которая выводит результат.
Формат ввода

В первой строке одно число n (0 ≤ n ≤ 2*105).

В следующих n строках заданы по два числа: цвет числа в ящике di и значение числа ai (-1018 ≤ di, ai ≤ 1018).

Гарантируется, что сумма чисел одного цвета не превышает 1018.
Формат вывода

Выведите в порядке возрастания номера цвета пары чисел, каждая в новой строке: номер цвета и сумму всех чисел данного цвета. 

In [ ]:
dic = {}

for _ in range(int(input())):
  a = list(map(int, input().split()))
  if a[0] not in dic.keys():
    dic[a[0]] = a[1]
  else:
    dic[a[0]] += a[1]
for key, i in sorted(dic.items()):
  print(key, i)


2
5 6
1 2
1 2
5 6


# C. Частотный анализ

Дан текст. Выведите все слова, встречающиеся в тексте, по одному на каждую строку. Слова должны быть отсортированы по убыванию их количества появления в тексте, а при одинаковой частоте появления — в лексикографическом порядке. Указание. После того, как вы создадите словарь всех слов, вам захочется отсортировать его по частоте встречаемости слова. Желаемого можно добиться, если создать список, элементами которого будут кортежи из двух элементов: частота встречаемости слова и само слово. Например, [(2, 'hi'), (1, 'what'), (3, 'is')]. Тогда стандартная сортировка будет сортировать список кортежей, при этом кортежи сравниваются по первому элементу, а если они равны — то по второму. Это почти то, что требуется в задаче. 

In [ ]:
n = input().split()
dic = {}
for i in n:
  dic[i]= dic.get(i, 0) +1
d = sorted(dic.items(),key=lambda para:(-para[1],para[0]))
for i in d:
  print(i[0])


hi hi what is your name my name is bond james bond my name is damme van damme claude van damme jean claude van damme
damme
is
name
van
bond
claude
hi
my
james
jean
what
your


In [ ]:
n= []

lines = open('input.txt')
for line in lines:
  n +=line.split()


dic = {}
for i in n:
  dic[i]= dic.get(i, 0) +1
d = sorted(dic.items(),key=lambda para:(-para[1],para[0]))
for i in d:
  print(i[0])

# B. Выборы в США


Как известно, в США президент выбирается не прямым голосованием, а путем двухуровневого голосования. Сначала проводятся выборы в каждом штате и определяется победитель выборов в данном штате. Затем проводятся государственные выборы: на этих выборах каждый штат имеет определенное число голосов — число выборщиков от этого штата. На практике, все выборщики от штата голосуют в соответствии с результами голосования внутри штата, то есть на заключительной стадии выборов в голосовании участвуют штаты, имеющие различное число голосов. Вам известно за кого проголосовал каждый штат и сколько голосов было отдано данным штатом. Подведите итоги выборов: для каждого из участника голосования определите число отданных за него голосов. 

In [ ]:
dic = {}
lines = open('input.txt')
for line in lines:
  a = line.split()
  if a[0] not in dic.keys():
    dic[a[0]] = int(a[1])
  else:
    dic[a[0]] += int(a[1])

for k, v in sorted(dic.items()):
  print(k, v)


gena 101
ivan 4
sergey 10000


# D. Выборы Государственной Думы



Статья 83 закона “О выборах депутатов Государственной Думы Федерального Собрания Российской Федерации” определяет следующий алгоритм пропорционального распределения мест в парламенте.

Необходимо распределить 450 мест между партиями, участвовавших в выборах. Сначала подсчитывается сумма голосов избирателей, поданных за каждую партию и подсчитывается сумма голосов, поданных за все партии. Эта сумма делится на 450, получается величина, называемая “первое избирательное частное” (смысл первого избирательного частного - это количество голосов избирателей, которое необходимо набрать для получения одного места в парламенте).

Далее каждая партия получает столько мест в парламенте, чему равна целая часть от деления числа голосов за данную партию на первое избирательное частное.

Если после первого раунда распределения мест сумма количества мест, отданных партиям, меньше 450, то оставшиеся места передаются по одному партиям, в порядке убывания дробной части частного от деления числа голосов за данную партию на первое избирательное частное. Если же для двух партий эти дробные части равны, то преимущество отдается той партии, которая получила большее число голосов.

Формат ввода

На вход программе подается список партий, участвовавших в выборах. Каждая строка входного файла содержит название партии (строка, возможно, содержащая пробелы), затем, через пробел, количество голосов, полученных данной партией – число, не превосходящее 108.

Формат вывода

Программа должна вывести названия всех партий и количество голосов в парламенте, полученных данной партией. Названия необходимо выводить в том же порядке, в котором они шли во входных данных. 

In [ ]:
import math
dic = {}
summ = 0
lines = open('input.txt')
for line in lines:
  v = line.split()
  st = ' '.join(v[:-1])
  summ += int(v[-1])
  if st not in dic.keys():
    dic[st] = int(v[-1])
  else:
    dic[st] += int(v[-1])

for k, v in dic.items():
  print(k, '{:.3f}'.format(v*450/summ))

Party 1 57.539
Party 2 109.058
Party 3 0.755
Party 4 109.641
Party 5 8.755
Party 6 5.947
Party 7 0.000
Party 8 0.002
Party 9 7.653
Party 10 150.650


In [ ]:
names = list()
votes = list()
sumVotes = 0
inFile = open('input.txt')
for line in inFile:
    line = line.split()
    partyName = ' '.join(line[:-1])
    partyVotes = int(line[-1])
    names.append(partyName)
    votes.append(partyVotes)
    sumVotes += partyVotes
inFile.close()
 
mandates = list()
fracPart = []
sumMandates = 0
for i in range(len(names)):
    partyMandates = (votes[i] * 450) / sumVotes
    print(partyMandates)
    sumMandates += int(partyMandates)
    print(sumMandates)
    mandates.append(int(partyMandates))
    fracPart.append(partyMandates - int(partyMandates))
    print(fracPart)
 
 
while sumMandates < 450:
    i = 0
    for j in range(1, len(fracPart)):
        if (
            (fracPart[j] > fracPart[i]) or  # если следующий остаток больше
            (fracPart[j] == fracPart[i] and votes[j] > votes[i])
                ):  # если пред и след остатки равны, но след голосов больше
            i = j
    mandates[i] += 1
    sumMandates += 1
    fracPart[i] = 0
 
for k in range(len(names)):
    print(names[k], mandates[k])

57.538972655251726
57
[0.538972655251726]
109.05787842087135
166
[0.538972655251726, 0.0578784208713472]
0.7550083889820998
166
[0.538972655251726, 0.0578784208713472, 0.7550083889820998]
109.64121823575817
275
[0.538972655251726, 0.0578784208713472, 0.7550083889820998, 0.6412182357581742]
8.755097278858655
283
[0.538972655251726, 0.0578784208713472, 0.7550083889820998, 0.6412182357581742, 0.7550972788586545]
5.946732741474905
288
[0.538972655251726, 0.0578784208713472, 0.7550083889820998, 0.6412182357581742, 0.7550972788586545, 0.9467327414749054]
0.0
288
[0.538972655251726, 0.0578784208713472, 0.7550083889820998, 0.6412182357581742, 0.7550972788586545, 0.9467327414749054, 0.0]
0.0016666851853909487
288
[0.538972655251726, 0.0578784208713472, 0.7550083889820998, 0.6412182357581742, 0.7550972788586545, 0.9467327414749054, 0.0, 0.0016666851853909487]
7.653418371315237
295
[0.538972655251726, 0.0578784208713472, 0.7550083889820998, 0.6412182357581742, 0.7550972788586545, 0.94673274147490

# Создание собственного множество!

In [ ]:
setsize = 10
myset = [[] for _ in range(setsize)]

def add(x):
  myset[x % setsize].append(x)

def find(x):
  for now in myset[x % setsize]:
    if now == x:
      return True
    return False

def delete(x):
  xlist = myset[x % setsize]
  for i in range(len(xlist)):
    if xlist[i] == x:
      xlist[i], xlist[len(xlist) - 1] = xlist[len(xlist) - 1], xlist[i]
      xlist.pop()
      return

# 5 Д/з

# A. Префиксные суммы

 В этой задаче вам нужно будет много раз отвечать на запрос «Найдите сумму чисел на отрезке в массиве».
Формат ввода
В первой строке записано два целых числа n и q (1≤n,q≤3⋅105) - размер массива и количество запросов.

Во второй строке записаны n целых чисел ai (1≤ai≤109) - сам массив.

Далее в q строках описаны запросы к массиву. Каждый запрос описывается двумя числами l, r (1≤l≤r≤n) - левой и правой границей отрезка, на котором нужно найти сумму.

Формат вывода
Для каждого запроса в отдельной строке выведите единственное число - сумму на соответствующем отрезке. 

Ввод
Вывод

4 10 

---


1 2 3 4

---


1 1
1 2
1 3
1 4
2 2
2 3
2 4
3 3
3 4
4 4

	

1
3
6
10
2
5
9
3
7
4


In [ ]:
n, q = map(int, input().split())
sp = list(map(int, input().split()))

for i in range(q):
  l, r = map(int, input().split())

  print(sum(sp[l-1: r]))

  

In [ ]:
n, q = map(int, input().split())
sp = list(map(int, input().split()))
otr = []
for i in range(q):
  l, r = map(int, input().split())
  otr.append([l,r])


for l, r in otr:
  s = l - 1
  summ = 0
  while s < r:
    summ +=sp[s]
    s += 1
  print(summ)


# B. Максимальная сумма

 В этой задаче вам требуется найти непустой отрезок массива с максимальной суммой.
Формат ввода
В первой строке входных данных записано единственное число n (1≤n≤3⋅105) -  размер массива.

Во второй строке записано n целых чисел ai (−109≤ai≤109) - сам массив.
Формат вывода
Выведите одно число - максимальную сумму на отрезке в данном массиве. 

In [ ]:
n = int(input())
sp = list(map(int, input().split()))
summ = sp[0]
lst_sum = [sp[0]]
for i in range(1, n):
  if summ + sp[i] > summ:
    summ += sp[i]
  else:
    lst_sum.append(summ)
    summ = sp[i]
  if sp[i] > max(lst_sum):
    lst_sum.append(sp[i])
lst_sum.append(summ)
print(max(lst_sum))

4
-1 -2 -3 -4
-1


In [ ]:
n = int(input())
sp = list(map(int, input().split()))
summ = sp[0]
lst_sum = [sp[0]]
for i in range(0, n-1):
  if summ + sp[i+1] > summ and summ >0:
    summ += sp[i+1]
  else:
    lst_sum.append(summ)
    summ = sp[i+1]
  if sp[i] > max(lst_sum):
    lst_sum.append(sp[i])
lst_sum.append(summ)
print(max(lst_sum))

7
805743163 -181176136 454376774 681211377 988713965 -599336611 -823748404 
2124302116


# C. Каждому по компьютеру



В новом учебном году на занятия в компьютерные классы Дворца Творчества Юных пришли учащиеся, которые были разбиты на N групп. В i-й группе оказалось Xi человек. Тут же перед директором встала серьезная проблема: как распределить группы по аудиториям. Во дворце имеется M ≥ N аудиторий, в j-й аудитории имеется Yj компьютеров. Для занятий необходимо, чтобы у каждого учащегося был компьютер и еще один компьютер был у преподавателя. Переносить компьютеры из одной аудитории в другую запрещается. Помогите директору!

Напишите программу, которая найдет, какое максимальное количество групп удастся одновременно распределить по аудиториям, чтобы всем учащимся в каждой группе хватило компьютеров, и при этом остался бы еще хотя бы один для учителя.
Формат ввода

На первой строке входного файла расположены числа N и M (1 ≤ N ≤ M ≤ 1000). На второй строке расположено N чисел — X1, …, XN (1 ≤ Xi ≤ 1000 для всех 1 ≤ i ≤ N). На третьей строке расположено M чисел Y1, ..., YM (1 ≤ Yi ≤ 1000 для всех 1 ≤ i ≤ M).

Формат вывода

Выведите на первой строке число P - количество групп, которые удастся распределить по аудиториям. На второй строке выведите распределение групп по аудиториям – N чисел, i-е число должно соответствовать номеру аудитории, в которой должна заниматься i-я группа. (Нумерация как групп, так и аудиторий, начинается с 1). Если i-я группа осталась без аудитории, i-е число должно быть равно 0. Если допустимых распределений несколько, выведите любое из них. 

In [ ]:
def readandnum():
  x = list(map(int, input().split()))
  for i in range(len(x)):
    x[i] = (x[i], i + 1)
  x.sort()
  return x

n, m = map(int, input().split())
x = readandnum()
y = readandnum()
cnt = 0
ynum = 0
ans = [0] * (n + 1)
for pupils, xnum in x:
  while ynum < len(y) and y[ynum][0] < pupils + 1:
    ynum += 1
  if ynum == len(y):
    break
  ans[xnum] = y[ynum][1]
  ynum += 1
  cnt += 1
print(cnt)
print(*ans[1:])

# D. Правильная, круглая, скобочная



Если из правильного арифметического выражения вычеркнуть всё, кроме круглых скобок, то получится правильная скобочная последовательность. Проверьте, является ли введённая строка правильной скобочной последовательностью.
Формат ввода

Вводится непустая строка, состоящая из открывающих и закрывающих круглых скобок. Длина строки не превосходит 100000
Формат вывода

Выведите YES если введённая строка является правильной скобочной последовательностью и NO иначе

In [ ]:
txt = input()
from itertools import accumulate
dct = {'(': 1, ')': -1}
tpl=tuple(accumulate((dct[i] for i in txt)))
print(('NO','YES')[min(tpl)>-1 and tpl[-1]==0])

In [ ]:
text = input()
if text[0] == ')' or text[-1] == '(':
    print('NO')
else:

  while (len(text) != 0) or (len(set(text)) > 1):
      text = text.replace('()','')
      if len(set(text)) == 1 or text == ')(':
          print('NO')
          break
  else: print('YES')


(())()
YES


# E. Сумма трёх

 Даны три массива целых чисел A,B,C и целое число S.

Найдите такие i,j,k, что Ai+Bj+Ck=S.
Формат ввода
На первой строке число S (1≤S≤109). Следующие три строки содержат описание массивов A,B,C в одинаковом формате: первое число задает длину n соответствующего массива (1≤n≤15000), затем заданы n целых чисел от 1 до 109 — сам массив.
Формат вывода
Если таких i,j,k не существует, выведите единственное число −1. Иначе выведите на одной строке три числа — i,j,k. Элементы массивов нумеруются с нуля. Если ответов несколько, выведите лексикографически минимальный. 

In [ ]:
S = int(input())
A = list(map(int, input().split()))
B = list(map(int, input().split()))
C = list(map(int, input().split()))
summ_ls = []

for i in range(1, len(A)):
  for j in range(1, len(B)):
    if S-(A[i]+B[j]) in C[1:]:
      summ_ls.append([i-1, j-1, C.index(S-(A[i]+B[j]))-1])
if len(summ_ls) == 0:
  print(-1)
else:
  summ_ls = sorted(summ_ls)
  print(*summ_ls[0])

6
2 2 1
2 2 2
2 2 4
0 0 -1


решение: проходимся по a и b с начала отсортированного списка, а c с конца (с самого большого) и сравниваем чтобы они были равны сумме и до этого такого не было.

In [ ]:
def readandenum():
  x = list(map(int, input().split()))[1:]
  for i in range(len(x)):
    x[i] = x[i], i
  x.sort()
  return x

s = int(input())
a = readandenum()
b = readandenum()
c = readandenum()
c.sort(key=lambda x: (x[0], -x[1]))
flag = False

for aval, apos in a:
  cpos = len(c) - 1
  for bval, bpos in b:
    while cpos > 0 and aval + bval + c[cpos][0] > s:
      cpos -= 1
    if aval + bval + c[cpos][0] == s and (not flag or (apos, bpos, cpos) < ans):
      ans  = apos, bpos, c[cpos][1]
      flag = True

if flag:
  print(*ans)
else:
  print(-1)

# A. Быстрый поиск в массиве

 Дан массив из N целых чисел. Все числа от −109 до 109.

Нужно уметь отвечать на запросы вида “Cколько чисел имеют значения отL доR?”.
Формат ввода
Число N (1≤N≤105). Далее N целых чисел.

Затем число запросов K (1≤K≤105).

Далее K пар чисел L,R (−109≤L≤R≤109) — собственно запросы.

Формат вывода
Выведите K чисел — ответы на запросы. 

In [ ]:
n = int(input())
list_n = sorted(list(map(int, input().split())))
k = int(input())
for _ in range(k):
  l, r = map(int, input().split())
  count = 0
  for i in list_n:
    if i >= l and i <= r:
      count += 1
  print(count, end=' ')

In [ ]:
n = int(input())
list_n = sorted(list(map(int, input().split())))
k = int(input())
dic = {}
for i in range(n):
  dic[list_n[i]] = list_n[i]
for _ in range(k):
  l, r = map(int, input().split())
  dic[l] = l
  dic[r] = r
  dic = sorted(dic)
  print(dic[l:r])

5
10 1 10 3 4
4
1 10
2 9
3 4
2 2
{1: 1, 3: 3, 4: 4, 10: 10, 2: 2, 9: 9}


# D. Вырубка леса



Фермер Николай нанял двух лесорубов: Дмитрия и Федора, чтобы вырубить лес, на месте которого должно быть кукурузное поле. В лесу растут X деревьев.

Дмитрий срубает по A деревьев в день, но каждый K-й день он отдыхает и не срубает ни одного дерева. Таким образом, Дмитрий отдыхает в K-й, 2K-й, 3K-й день, и т.д.

Федор срубает по B деревьев в день, но каждый M-й день он отдыхает и не срубает ни одного дерева. Таким образом, Федор отдыхает в M-й, 2M-й, 3M-й день, и т.д.

Лесорубы работают параллельно и, таким образом, в дни, когда никто из них не отдыхает, они срубают A + B деревьев, в дни, когда отдыхает только Федор — A деревьев, а в дни, когда отдыхает только Дмитрий — B деревьев. В дни, когда оба лесоруба отдыхают, ни одно дерево не срубается.

Фермер Николай хочет понять, за сколько дней лесорубы срубят все деревья, и он сможет засеять кукурузное поле.

Требуется написать программу, которая по заданным целым числам A, K, B, M и X определяет, за сколько дней все деревья в лесу будут вырублены.
Формат ввода

Входной файл содержит пять целых чисел, разделенных пробелами: A, K, B, M и X (1 ≤ A, B ≤ 109 , 2 ≤ K, M ≤ 1018, 1 ≤ X ≤ 1018).
Формат вывода

Выходной файл должен содержать одно целое число — искомое количество дней. 

In [ ]:
a, k, b, m, x = map(int, input().split())

l = 0
r = x * 2 // a + 1
while l < r:
  days = (l + r) // 2
  hd = days // k
  hf = days // m
  lumber = (days - hd) * a + (days - hf) * b
  if lumber < x:
    l = days + 1
  else:
    r = days
print(l)

# Лекция 4: «Сортировка подсчётом»

**сортировака подсчетом**

In [ ]:
def countsort(seq):
  minval = min(seq)
  maxval = max(seq)
  k = (maxval - minval + 1)
  count = [0] * k

  for now in seq:
    count[now - minval] += 1
  nowpos = 0
  for val in range(0, k):
    for i in range(count[val]):
      seq[nowpos] = val + minval
      nowpos += 1

# Лекция 5: «Префиксные суммы и два указателя»

Массив можно построить за O(N): prefsum[i] = prefsum[i-1] + nums[i-1]

Не забыть про отличие размера!

Переполнение!

*Ответ на запрос суммы на отрезке*

Ответ за O(1): sum(L, R) = prefsum[R] - prefsum[L]

Найдем sum(2, 5) = prefsum[5] - prefsum[2] = 21 - 8 = 13

In [ ]:
def makeprefixsum(nums):
  prefixsum = [0] * (len(nums) + 1)
  for i in range(1, len(nums) + 1):
    prefixsum[i] = prefixsum[i - 1] + nums[i - 1]
  return prefixsum

def rsq(prefixsum, l, r):
  returnprefixsum[r] - prefixsum[l]

Задача №1 

Дана последовательность чисел длинной N и M запросов 

Запрасы: "Сколько нулей на полуинтервале [L, R)"

Решение за O(N + M)

Для каждого префикса посчитаем количество нулей на нем (prefixzeroes). Тогда ответ на запрос на полуинтервале [L, R): prefixzeroes[R] - prefixzeroes[L]

nums 1 0 1 1 0 0 1
pref 0 0 1 1 1 2 3 3

In [ ]:
def makeprefixzeroes(nums):
  prefixzeroes = [0] * (len(nums) + 1)
  for i in range(1, len(nums) + 1):
    if nums[i - 1] == 0:
      prefixzeroes[i] = prefixzeroes[i - 1] + 1
    else:
      prefixzeroes[i] = prefixzeroes[i - 1]
  return prefixzeroes
def countzeroes(prefixzeroes, l, r):
  return prefixzeroes[r] - prefixzeroes[l]

Задача №2

Дана последовательность чисел длинной N

Необходимо найти количество отрезков с нулевой суммой 


Решение за O(N**2)

Переберем начало и будем двагать конец, суммируя элементы

In [ ]:
def countzerosumranges(nums):
  cntranges = 0
  for i in range(len(nums)): #L граница
    rangesum = 0
    for j in range(i, len(nums)): #R граница
        rangesum += nums[j]
        if rangesum == 0:
          cntranges += 1
  return cntranges

Решение за O(N)

Насчитаем префиксные суммы. Одинаковый преффиксные суммы означают, что сумма на отрезке с началом и концом в позициях, на которых достигаются одинаковые префиксные суммы, будет равна нулю

In [ ]:
def countprefixsums(nums):
  prefixsumbyvalue = {0: 1}
  nowsum = 0
  for now in nums:
    nowsum += now
    if nowsum not in prefixsumbyvalue:
      prefixsumbyvalue[nowsum] = 0
    prefixsumbyvalue[nowsum] += 1
  return prefixsumbyvalue

def countzerosumranges(prefixsumbyvalue):
  cntranges = 0
  for nowsum in prefixsumbyvalue:
    cntsum = prefixsumbyvalue[nowsum]
    cntranges += cntsum * (cntsum - 1) // 2
  return cntranges

***Два указателя ***


Задача №3

Дана отсортированная последовательность чисел длиной N и число K

Необходимо найти количество пар чисел А, В, таких что В - А > K

Решение за O(N):

Возьмем наименьшее число и найдем для него первое подходящее большее. Все еще большие число точно подходят. Возьмем в качестве меньшего числа следующее, а указатель первого подходящего большего будем двигать начиная с той позиции, где он находится чейчас

In [ ]:
def cntpairswitdiff(sortednums, k):
  cntpairs = 0
  last = 0
  for first in range(len(fortednums)):
    while last < len(sortednums) and sortednums[last] - sortednums[first] <= k:
      last += 1
    cntpairs += len(sortednums) - last 
  return cntpairs

Задача MERGE

Даны две отсортированные последовательности чисел (длиной N и M соответственно)
Необходимо слить их в одну отсортированную последовательность.

Решение

Поставим два указателя на начало каждой из последовательностей. Выребем тот, который указывает на меньшее число, запишем это число в результат и сдвинем указатель

*Более близкая к идеалу реализация *

In [ ]:
def merge(nums1, nums2):
  merged = [0] * (len(nums1) + len(nums2))
  first1 = first2 = 0
  for k in range(len(nums1) + len(nums2)):
    if first1 != len(nums1) and (first2 == len(nums2) or nums1[first1] <= nums2[first2]):
      merged[k] = nums1[first1]
      first1 += 1
    else:
      merged[k] = nums2[first2]
      first2 += 1
  return merged

# Лекция №6: Бинарный поиск

Левый бинарный поиск

In [ ]:
def lbinsearch(l, r, check, checkparams):
  while l < r:
    m = (l + r) // 2
    if check(m, checkparams):
      r = m
    else:
      l = m + 1
  return l

Правый бинарный поиск

In [ ]:
def rbinsearch(l, r, check, checkparams):
  while l < r:
    m = (l + r + 1) // 2
    if check(m, checkparams):
      l = m
    else:
      r = m - 1
  return l

Задача №1

В совет входят учетеля, родители и ученики, причем родителей больжно быть не менее трети от общего числа членов совета. В настоящий момент в совет входит N человек, из ник K родителей.
Определите сколько родителей нужно дополнительно ввести в совет, чтобы их число стало составлять не менее трети от число членов совета.

Решение:
Бедем искать мин кол родителей, которых нужно добавить, бинарным поиком. Не забываем, что новые родители добавляются к общему числу членов совета. Не спользуем деление. L = 0, R = N

In [ ]:
def lbinsearch(l, r, check, checkparams):
  while l < r:
    m = (l + r) // 2
    if check(m, checkparams):
      r = m
    else:
      l = m + 1
  return l

def checkendownment(m, params):
  n, k = params
  return (k + m) * 3 >= n + m

Задача №2

Юра решает задачи. Выбрал N задача. В первый день решил K задач, а каждый следующий день решает на одну задачу больше, чем в предыдущий день.
Определите, сколько дней уйдет на подготовку.

Решение:
Будем искать мин кол дней, достаточное для решения не менее N задач, бинарным поиском. Нам понадобится формула арифметической прогрессии L = 0, R = N

In [ ]:
def lbinsearch(l, r, check, checkparams):
  while l < r:
    m = (l + r) // 2
    if check(m, checkparams):
      r = m
    else:
      l = m + 1
  return l

def checkproblemcount(days, params):
  n, k = params
  return (k + (k + days - 1)) * days // 2 >= n

Задача 

Задана отсортированная по неубыванию последовательность из N чисел и число X. 
Необходимо определить индекс первого числа в последовательности, которое больше либо равно X. Если такого числа нет, вернуть число N

Решение: воспользуемся левым бинпоиском для поиска первого подходящего числа. В случае, если бинпоиск сошелся к ччислу, меньшему X вернем N

In [ ]:
def checkisge(index, params):
  seq, x = parems
  return seq[index] >= x

def findfirstge(seq, x):
  ans = lbinsearch(0, len(seq) - 1, checkisge, (seq, x))
  if seq[ans] < x:
    return len(seq)
  return ans

Задача

Задана отсортированная по неубыванию последовательность из N чисел и число X. Необходимо определить сколько раз число X входит в последовательность

Решение: найдем одним левым бинпоиском первое число большее либо равное X, а вторым левым бинпоиском - число строго большее X. Разность индексов и будет количеством вхождений

In [ ]:
def checkisge(index, params):
  seq, x = parems
  return seq[index] >= x

def checkisgt(index, params):
  seq, x = parems
  return seq[index] > x

def findfirst(seq, x):
  ans = lbinsearch(0, len(seq) - 1, checkisge, (seq, x))
  if not check(ans, (seq, x):
    return len(seq)
  return ans

def countx(seq, x):
  indexgt = findfirst(seq, x, checkisgt)
  indexge = findfirst(seq, x, checkisge)
  return indexgt - indexge

# Лекция 7: «Сортировка событий»

Задача №1 

сайт поситело N человек, для каждого известно время входа на сайт In и время выхода с сайта Out. Считается, что человек был на сайте с момента In по Out включительно.
Определите, какое мак кол человек было на сайте одновременно. 

Решение: Создадим на каждого чедовека два события: вход и выход. Каждое событие - пара, в которой первое число - время, а второе - тип события.
Событие "вход на сайт" должно происходить раньше "выхода с сайта".

In [ ]:
def maxvisitorsonline(n, tin, tout):
  events = []
  for i in range(n):
    events.append((tin[i], -1))
    events.append((tout[i], 1))
  events.sort()
  online = 0
  maxonline = 0
  for event in events:
    if event[1] == -1:
      online += 1
    else:
      online -= 1
    maxonline = max(online, maxonline)
  return maxonline

Задача №2

сайт поситело N человек, для каждого известно время входа на сайт In и время выхода с сайта Out. Считается, что человек был на сайте с момента In по Out включительно. Определите, какое суммарное время на сайте был хотя бы один человек.


Решение: если мы пришли в событие с положительным счетчиком кол людей, то между этим и предыдущим событием на сайте кто-то был. Прибавим к ответу время между текущим и предыдущим событием.

In [ ]:
def timewithvisitors(n, tin, tout):
  events = []
  for i in range(n):
    events.append((tin[i], -1))
    events.append((tout[i], 1))
  events.sort()
  online = 0
  notemptytime = 0
  for i in range(len(events)):
    if online > 0:
      notemptytime += events[i][0] - events[i - 1][0]
    if events[i][1] == -1:
      online += 1
    else:
      online -= 1
  return notemptytime

Задача №3

сайт поситело N человек, для каждого известно время входа на сайт In и время выхода с сайта Out. Считается, что человек был на сайте с момента In по Out включительно. Начальник заходил на сайт M раз в моменты времени Boss и смотрел, сколько людей сейчас онлайн. Посещения сайта начальником упорядочены по позрастанию времени. 
Определите, какие показания счетчика людей онлайн увидел начальник.

Решение: создадим дретий тип события - "вход начальника" и при наступлении такого события будем сохранять текущее значение счетчика посетителей 

In [ ]:
def bosscounters(n, tin, tout, m tboss):
  events = []
  for i in range(n):
    events.append((tin[i], -1))
    events.append((tout[i], 1))
  for i in range(m):
    events.append((tboss[i], 0))
  events.sort()
  online = 0
  bossans = 0
  for i in range(len(events)):
    if events[i][1] == -1:
      online += 1
    elif events[i][1] == 1:
      online -= 1
    else:
      bossans.append(online)
  return bossans

Задача №4

На парковке N мест(нумерация от 1 до N). За день приезжало туда M авто, при этом некоторые из них длинные и занимали несоклько подряд идущих мест. Для каждого авто известно время приезда и отъезда, а также два числа - к какого по какое парковочное места он занимал. Необходимо определить, был ли моментЮ в который были заняты все парковочные места.

Решение: события - приезд и отъезд авто (причем отъезд должен происходить раньше). Будем поддерживать количество занятых мест и если после очередного события счетчик равен N, то такие мометны были.

In [ ]:
def isparkfull(cars, n):
  events = []
  for car in cars:
    timein, timeout, placefrom, placeto = car
    events.append((timein, 1, placeto - placefrom + 1))
    events.append(timeout, -1, placeto - placefrom + 1))
  events.sort()
  occupied = 0
  for i in range(len(events)):
    if events[i][1] == -1:
      occupied -= events[i][2]
    elif events[i][1] == 1:
      occupied += events[i][2]
    if occupied == n:
      return True
  return False

Задача №5

На парковке N мест(нумерация от 1 до N). За день приезжало туда M авто, при этом некоторые из них длинные и занимали несоклько подряд идущих мест. Для каждого авто известно время приезда и отъезда, а также два числа - к какого по какое парковочное места он занимал. 
Необходимо определить, был лм момент, в который были заняты все парковочные места и определить мин кол авто, которое заняло все места. Если такого момента не было - вернуть M + 1

Решение: добавим еще один счетчик на кол авто и будем обновлять мин кол авто когда заняты все места

In [ ]:
def mincfrsonfullpfrking(cars, n):
  events = []
  for car in cars:
    timein, timeout, placefrom, placeto = car
    events.append((timein, 1, placeto - placefrom + 1))
    events.append(timeout, -1, placeto - placefrom + 1))
  events.sort()
  occupied = 0
  nowcars = 0
  mincars = len(cars) + 1
  for i in range(len(events)):
    if events[i][1] == -1:
      occupied -= events[i][2]
      nowcars -= 1
    elif events[i][1] == 1:
      occupied += events[i][2]
      nowcars += 1
    if occupied == n:
      mincars = min(mincars, nowcars)
  return mincars

# Лекция 8: «Деревья»

**Менеджмент памяти **

У нас есть заранее неизвестное кол структур с двуммя ссылками на другие структуры

Мы знаем заранее, какое макс кол структур может существовать одновременно 

Хотим научить выделять и освобождать память

Код менеджера памяти:

In [ ]:
def initmemory(maxn):
  memory = []
  for i in range(maxn):
    memory.append([0, i + 1, 0])
  return [memory, 0]

def newnode(memstruct):
  memory, firstfree = memstruct
  memstruct[1] = memory[firstfree][1]
  return firstfree

def delnode(memstruct, index):
  memory, firstfree = memstruct
  memory[index][1] = firstfree
  memstruct[1] = index

Бинарное дерево поиска 

У каждого узла ключ и два сына - левый и правый

В левом поддереве ключи меньше, а в правом - больше

Если ключи поступают в случайном порядке, то глубина дерева будет O(logN)

Поиск в дереве:

In [ ]:
def find(memstruct, root, x):
  key = memstruct[0][root][0]
  if x == key:
    return root
  elif x < key:
    left = memstruct[0][root][1]
    if left == -1:
      return -1
    else:
      return find(memstruct, left, x)
  elif x > key:
    right = memstruct[0][root][2]
    if right == -1:
      return -1
    else:
      return find(memstruct, right, x)

Добавление в дерево:

In [ ]:
def createandfillnode(memstruct, key):
  index = newnode(memstruct)
  memstruct[0][index][0] = key
  memstruct[0][index][1] = -1
  memstruct[0][index][2] = -1
  return index

def add(memstruct, root, x):
  key = memstruct[0][root][0]
  elif x < key:
    left = memstruct[0][root][1]
    if left == -1:
      memstruct[0][root][1] = createandfillnode(memstruct, x)
    else:
      return add(memstruct, left, x)
  elif x > key:
    right = memstruct[0][root][2]
    if right == -1:
      memstruct[0][root][2] = createandfillnode(memstruct, x)
    else:
      return add(memstruct, right, x)

Инидиализация дерева:

In [ ]:
memstruct = initmemory(20)
root = createandfillnode(memstruct, 8)
add(memstruct, root, 10)
add(memstruct, root, 9)
add(memstruct, root, 14)
add(memstruct, root, 13)
add(memstruct, root, 4)